p2p/async-buffer/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+
+        BufferSinkReady {
+            sink: self,
+            size_needed,
+        }
+    }
+
+    /// Attempts to add an item to the buffer returning an error if there is not enough capacity or
+    /// the [`BufferStream`] was dropped.
+    pub fn try_send(&mut self, item: T, size_needed: usize) -> Result<(), BufferError> {
+        let size_needed = min(self.max_item_weight, size_needed);
+
+        if self.capacity.load(Ordering::Acquire) < size_needed {
+            return Err(BufferError::NotEnoughCapacity);
+        }
+
+        let prev_size = self.capacity.fetch_sub(size_needed, Ordering::AcqRel);
+
+        // make sure we haven't wrapped the capacity around.
+        assert!(prev_size >= size_needed);
+
+        self.queue
+            .unbounded_send((item, size_needed))
+            .map_err(|_| BufferError::Disconnected)?;
+
+        Ok(())
+    }
+
+    /// Waits for capacity in the buffer and then sends the item.
+    pub fn send(&mut self, item: T, size_needed: usize) -> BufferSinkSend<'_, T> {
+        BufferSinkSend {
+            ready: self.ready(size_needed),
+            item: Some(item),
+        }
+    }
+}
+
+/// A [`Future`] for adding an item to the buffer.
+#[pin_project::pin_project]
+pub struct BufferSinkSend<'a, T> {
+    /// A future that resolves when the channel has capacity.
+    #[pin]
+    ready: BufferSinkReady<'a, T>,
+    /// The item to send.
+    item: Option<T>,
+}
+
+impl<'a, T> Future for BufferSinkSend<'a, T> {
+    type Output = Result<(), BufferError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        let size_needed = this.ready.size_needed;
+
+        this.ready.as_mut().poll(cx).map(|_| {
+            this.ready
+                .sink
+                .try_send(this.item.take().unwrap(), size_needed)
+        })
+    }
+}
+
+/// A [`Future`] for waiting for capacity in the buffer.
+pub struct BufferSinkReady<'a, T> {
+    /// The sink side of the buffer.
+    sink: &'a mut BufferAppender<T>,
+    /// The capacity needed.
+    ///
+    /// This future will wait forever if this is higher than the total availability of the buffer.
+    size_needed: usize,
+}
+
+impl<'a, T> Future for BufferSinkReady<'a, T> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Check before setting the waker just in case it has capacity now,
+        if self.sink.capacity.load(Ordering::Acquire) >= self.size_needed {
+            return Poll::Ready(());
+        }
+
+        // set the waker
+        self.sink.sink_waker.register(cx.waker());
+
+        // check the capacity again to avoid a race condition that would result in lost notifications.
+        if self.sink.capacity.load(Ordering::Acquire) >= self.size_needed {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
```

Comment:
I'm not sure if this is correct in the context of the previous atomics (this function could return `Ready` after 118 but before 122).

Reply:
Ignore since SPSC.